grpc_adapter.go

```diff
@@ -0,0 +1,668 @@
+	contentType := grpcContentTypePrefix + enc
+	setHeaderCanonical(header, headerContentType, contentType)
+}
+
+type grpcWebResponseWriter struct {
+	http.ResponseWriter
+	typ, enc string
+
+	seenHeaders map[string]bool
+	wroteHeader bool
+}
+
+func newGRPCWebResponseWriter(responseWriter http.ResponseWriter, typ, enc string) *grpcWebResponseWriter {
+	return &grpcWebResponseWriter{
+		ResponseWriter: responseWriter,
+		typ:            typ,
+		enc:            enc,
+	}
+}
+
+func (w *grpcWebResponseWriter) writeHeaders() {
+	header := w.Header()
+
+	isTrailer := map[string]bool{
+		// Force GRPC status values to be trailers.
+		grpcHeaderStatus:  true,
+		grpcHeaderDetails: true,
+		grpcHeaderMessage: true,
+	}
+	for _, key := range strings.Split(header.Get(headerTrailer), ",") {
+		key = http.CanonicalHeaderKey(key)
+		isTrailer[key] = true
+	}
+
+	keys := make(map[string]bool, len(header))
+	for key := range header {
+		if strings.HasPrefix(key, http.TrailerPrefix) {
+			continue
+		}
+		canonicalKey := http.CanonicalHeaderKey(key)
+		if key != canonicalKey {
+			header[canonicalKey] = header[key]
+			delete(header, key)
+			key = canonicalKey
+		}
+		if isTrailer[key] {
+			continue
+		}
+		keys[key] = true
+	}
+
+	setHeaderCanonical(header, headerContentType, w.typ+"+"+w.enc)
+
+	w.seenHeaders = keys
+	w.wroteHeader = true
+}
+
+func (w *grpcWebResponseWriter) Write(b []byte) (int, error) {
+	if !w.wroteHeader {
+		w.writeHeaders()
+	}
+	return w.ResponseWriter.Write(b)
+}
+
+func (w *grpcWebResponseWriter) WriteHeader(code int) {
+	if !w.wroteHeader {
+		w.writeHeaders()
+	}
+	w.ResponseWriter.WriteHeader(code)
+}
+
+func (w *grpcWebResponseWriter) flushWithTrailers() {
+	if !w.wroteHeader {
+		w.writeHeaders()
+	}
+	if err := w.writeTrailer(); err != nil {
+		return // nothing
+	}
+	flushResponseWriter(w.ResponseWriter)
+}
+
+func (w *grpcWebResponseWriter) Flush() {
+	flushResponseWriter(w.ResponseWriter)
+}
+
+func (w *grpcWebResponseWriter) writeTrailer() error {
+	header := w.Header()
+	trailer := make(http.Header, len(header)-len(w.seenHeaders))
+	for key, values := range header {
+		if w.seenHeaders[key] {
+			continue
+		}
+		key = strings.TrimPrefix(key, http.TrailerPrefix)
+		// https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-WEB.md#protocol-differences-vs-grpc-over-http2
+		trailer[strings.ToLower(key)] = values
+	}
+
+	var buf bytes.Buffer
+	if err := trailer.Write(&buf); err != nil {
+		return err
+	}
+
+	head := [5]byte{1 << 7, 0, 0, 0, 0} // MSB=1 indicates this is a trailer data frame.
+	binary.BigEndian.PutUint32(head[1:5], uint32(buf.Len()))
+	if _, err := w.ResponseWriter.Write(head[:]); err != nil {
+		return err
+	}
+	if _, err := w.ResponseWriter.Write(buf.Bytes()); err != nil {
+		return err
+	}
+	return nil
+}
+
+// bufferedEnvelopeReader buffers a message to wrap it in an envelope.
+type bufferedEnvelopeReader struct {
+	io.ReadCloser
+
+	config     *grpcAdapterConfig
+	buf        bytes.Buffer
+	buffered   bool
```

Comment:
On first read, this was a bit confusing. I understand it now, but I think `initialized` might be a better word since the buffering is an initialization step.

Reply:
I've renamed this to:
```go
	buf            bytes.Buffer
	isInitialized  bool
	initializedErr error
	isCompressed   bool
```